test/reducer-e2e/reducer_test.go

```diff
@@ -0,0 +1,77 @@
+package reducer_e2e
+
+import (
+	dfv1 "github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	. "github.com/numaproj/numaflow/test/fixtures"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+)
+
+type ReducerSuite struct {
+	E2ESuite
+}
+
+func (s *ReducerSuite) TestSimpleReducerPipelineFailOver() {
+	w := s.Given().Pipeline("@testdata/simple-reduce-pipeline.yaml").
+		When().
+		CreatePipelineAndWait()
+	defer w.DeletePipelineAndWait()
+
+	pipelineName := "even-odd-sum"
+
+	w.Expect().
+		VertexPodsRunning().DaemonPodsRunning().
+		VertexPodLogContains("in", LogSourceVertexStarted).
+		VertexPodLogContains("atoi", LogUDFVertexStarted, PodLogCheckOptionWithContainer("numa")).
+		VertexPodLogContains("sink", LogSinkVertexStarted).
+		DaemonPodLogContains(pipelineName, LogDaemonStarted)
+
+	defer w.VertexPodPortForward("in", 8443, dfv1.VertexHTTPSPort).
+		TerminateAllPodPortForwards()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		startTime := int(time.Unix(1000, 0).UnixMilli())
+		for i := 1; i <= 100; i++ {
+			for j := 0; j < 1; j++ {
+				eventTime := startTime + (i * 1000)
+				HTTPExpect(s.T(), "https://localhost:8443").POST("/vertices/in").WithHeader("X-Numaflow-Event-Time", strconv.Itoa(eventTime)).WithBytes([]byte("1")).
+					Expect().
+					Status(204)
+
+				HTTPExpect(s.T(), "https://localhost:8443").POST("/vertices/in").WithHeader("X-Numaflow-Event-Time", strconv.Itoa(eventTime)).WithBytes([]byte("2")).
+					Expect().
+					Status(204)
+
+				HTTPExpect(s.T(), "https://localhost:8443").POST("/vertices/in").WithHeader("X-Numaflow-Event-Time", strconv.Itoa(eventTime)).WithBytes([]byte("3")).
+					Expect().
+					Status(204)
+			}
+		}
+
+		w.Expect().VertexPodLogContains("sink", "Payload -  120")
+		w.Expect().VertexPodLogContains("sink", "Payload -  240")
+	}()
+
+	// Kill one of the pod during processing to trigger failover.
+	block := func(t *testing.T, output string, err error) {
```

Comment:
should we initiate a kill at a very specific point eg, at `i=x && j=y`? this will ensure that we do not kill at random points which could end up being flaky.

Reply:
This is to mimic that failure (or pod migration) can happen at any moment during pipeline execution and still get the correct result. If there is any flakiness due to this, I think we should investigate and check what is the cause.

Reply:
such random testing shouldn't be done in e2e tests. E2E tests should be predictable. Such random tests can be labeled, perhaps under Fuzz tests or something.

There are three major of pod migration scenarios
* right before processing
* during processing
* right after processing

Reply:
Do you want me to rename the the test or create a new folder for fuzz tests and put it under it? I can add the other two scenarios, for `right before processing` and `right after processing` I didn't add previously as the behavior shouldn't differ with or without WAL implementation.

Reply:
Let's not do fuzz testing now. Instead, let's make this E2E test complete by adding the 3 scenarios I have mentioned.

Reply:
@xdevxy - let's do following:

1. Create the pipeline, expect both reduce pods running;
2. Kill them both;
3. Start ingesting messages, pause when half is done;
4. Expect both reduce pods running, and kill them again;
5. Finish ingesting;
6. Expect the results,
7. Kill them again, and no result changes expected.